[WIP] support for tutorials/guides

errordeveloper opened the pull request:
@awh PTAL, it needs a lot of cleanup, but I've got a rough idea what to do next
@lukemarsden FYI – I've made some progress on getting something basic working

errordeveloper commented on json.go:
```diff
@@ -22,6 +22,20 @@ type Document struct {
 	Status    string `json:"status"`
 }
 
+type Guide struct {
```
I'm actually not even using this yet, not sure if I'll need it really...

awh replied:
To use it, you would have to generalise most of the shared stuff (e.g. `parse.go`, `rewrite.go`, `api.go`) - does that sound appealing in Go? :wink:

errordeveloper replied:
I guess this is the biggest thing I wanted to discuss – the difference seems very very minor. I could potentially just insert a few `if`'s into the existing code, and may be rename a few things, like `Document` could become something like `DocumentOrGuide` or `AnyDocument`, or is a guide just a `Document{Product: "tutorials", Tag: "current"}`..?

awh commented:
One other thing - no doubt the cleanup will address this, but lets have nicely formed descriptive commits please!

errordeveloper commented on cmd/wordepress/cmd/publish-guide.go:
```diff
@@ -0,0 +1,121 @@
+)
+
+var publishGuideCmd = &cobra.Command{
+	Use:   "publish-guide",
```
Sure, this makes most sense, as we discussed before, I will make it so.

errordeveloper commented on plugin/wordepress/wordepress.php:
```diff
@@ -2,88 +2,99 @@
+add_action( 'init', function () { wordepress_init_post_type ( 'documentation' ); }, 100);
+
+// Register the `guide` post type with the REST API
+add_action( 'init', function() { wordepress_init_post_type ( 'tutorials' ); }, 100);
```
Yeah, also panned to clean this up – the reason we have to use word `tutorial` is because there are `guides` in WP already. We need some kind of migration plan and I have no idea what's best, a new name seemed reasonable.

errordeveloper commented on plugin/wordepress/wordepress.php:
```diff
@@ -2,88 +2,99 @@
 
     add_rewrite_rule(
-		'docs/([^/]+)/([^/]+)/([^/]+)/([^/]+)',
+        'docs/([^/]+)/([^/]+)/([^/]+)/([^/]+)',
```
> Is this whitespace change necessary?

Thanks for asking! Of course it's not :)

errordeveloper commented:
@awh thanks for looking at this, may be we can discuss a bit more f2f or on Slack 👍

errordeveloper commented:
@awh I've done the renaming that we discussed in https://github.com/weaveworks/wordepress/pull/34/commits/8236393dc5ff080a2993d1e9e3d09af23e573d45, PTAL and LMK what you think. For example, I've ended up renaming `PostDocument()` to `Post()` as `PostPost()` didn't seem right and other CRUD methods followed the same pattern – happy to discuss alternatives :)

errordeveloper commented:
@awh I've moved the code around and added sub-commands as we discussed. I think another review would be great.  By the way, I wouldn't mind to cherry-pick the commit with renaming into a separate PR, if you are happy with that one as is.

errordeveloper commented:
This now implement a complete set of features we need for guides, including templates.

errordeveloper commented:
@awh I've renamed post type slug to `tutorials-post-type` and  made rewrite rules use `/tutorials`, however URL in the browser still end-up `/tutorials-post-type/tutorials-latest-foo/` and not `/tutorials/latest/part-foo/`.

I can see this in pre-defined redirects:
![](https://www.dropbox.com/s/eunt7mxj1vde7f0/Screenshot%202017-01-10%2016.29.27.png?dl=1)
Could that be it?

errordeveloper commented:
@awh would you like me to make this easier to review by cherry-picking 0e4b021 into another PR?

errordeveloper commented:
@awh if you want to test it, you can use https://github.com/weaveworks/guides/tree/reautomate.

awh commented:
@errordeveloper two problems so far:

1. The 'tutorials' CPT is not included in the toolset exports so I can't test tutorial publishing
1. The changes to the markdown rendering aren't compatible with the existing documentation - you need to make sure we can still publish the weave and scope docs successfully

awh commented:
@errordeveloper overall this looks in good shape; the only thing holding it up are the issues above. Once I can test the tutorial publishing locally I can try to get the redirect working properly...

errordeveloper commented:
> The changes to the markdown rendering aren't compatible with the existing documentation

@awh could you please elaborate on this?

awh commented:
@errordeveloper I don't recall the precise details, and it's not easy for me to replicate right now, but if you try to publish the weave net docs using `wordepress` from your branch it you'll see what I mean - it errors out pretty quickly.

Having said that, given that we're getting rid of Wordpress perhaps it's not worth fixing - we could just keep using the `v1.0.0` tag for publishing the net & scope docs until that happens, wdyt?

errordeveloper commented:
> Having said that, given that we're getting rid of Wordpress perhaps it's not worth fixing - we could just keep using the v1.0.0 tag for publishing the net & scope docs until that happens, wdyt?

Yeah, sounds plausible. Let's see if we can get PHP changes to work with both versions... or we could just rename the new version in an ad-hoc fashion?
